### ISSUE TITLE
panic "Base types cannot be unioned"

### ISSUE THREAD

**Issue by nlewycky** [egraphs-good/egglog#235]
This code when input to the web demo:

```
(datatype Bool)
(declare True Bool)
(declare False Bool)

(function bool-to-i64 (Bool) i64)
(rewrite (bool-to-i64 False) 0)
(rewrite (bool-to-i64 True) 1)
```

causes a panic! with the message "Base types cannot be unioned".

**Comment by oflatt** [comment]
Hi- this is the intended behavior.
The reason for this is that egglog disallows unioning primitive types like `i64`. If you want something that can be made equal to other things, consider wrapping your integers in a constructor like `Num` in the `math.egg` example.

The `rewrite` command is short-hand for the following code:
```
(rule ((= lhs (bool-to-i64 False)))
         ((union lhs 0)))
```
Which uses `union` to make things equal.

**Comment by nlewycky** [comment]
Thanks for the explanation!

Perhaps an error message instead of a `panic!`?

**Comment by oflatt** [comment]
Yes, an error message with a helpful explanation would be much better! Keeping the issue open

**Comment by saulshanabrook** [comment]
Another option is to use the `set` command instead of `rewrite` which works for functions that return primitives, unlike rewrite:

```
(datatype Bool)
(declare True Bool)
(declare False Bool)

(function bool-to-i64 (Bool) i64)
(set (bool-to-i64 False) 0)
(set (bool-to-i64 True) 1)
```

The way I understand it, for any unique set of inputs, functions that return primtives can either be unset or have one value they are set to. Whereas functions that return non primitives can be unioned with any number of other values of the same sort.

**Comment by Alex-Fischman** [comment]
I think this was fixed in #461.
